fastestimator/op/tensorop/model/update.py

```diff
@@ -74,22 +107,130 @@ def fe_retain_graph(self, retain: Optional[bool] = None) -> Optional[bool]:
         return self.retain_graph
 
     def forward(self, data: Union[Tensor, List[Tensor]], state: Dict[str, Any]) -> None:
-        if not state["warmup"]:
-            if self.gradients is None:
-                if self.weight_decay:
-                    data = data + tf.reduce_sum(self.model.losses)
-                update_model(self.model,
-                             loss=data,
-                             tape=state['tape'],
-                             retain_graph=self.retain_graph,
-                             scaler=state["scaler"],
-                             defer=self.defer,
-                             deferred=state["deferred"])
-            else:
-                update_model(self.model,
-                             gradients=data,
-                             tape=state['tape'],
-                             retain_graph=self.retain_graph,
-                             scaler=state["scaler"],
+        if state["warmup"]:
+            return
+
+        if self.gradients is None:  # data is loss
+            loss = self._loss_preprocess(data, state)
+            gradients = self._get_gradient(loss, state)
+            gradients = self._gradient_postprocess(gradients, state)
+
+        else:  # data is gradients
+            gradients = data
+
+        if self.merge_grad > 1:
+            self._merge_grad_update(gradients, state)
+        else:
+            update_model(model=self.model,
+                         gradients=gradients,
+                         tape=state["tape"],
+                         defer=self.defer,
+                         deferred=state["deferred"])
+
+    def _loss_preprocess(self, loss, state):
+        """Loss preprocess for multi-GPU and mixed-precision training
+        """
+        if self.weight_decay:
+            loss = loss + tf.reduce_sum(self.model.losses)
+        loss = reduce_mean(loss)
+
+        if self.framework == "tf":
+            # scale up loss for mixed precision training to avoid underflow
+            if self.model.mixed_precision:
+                loss = self.model.current_optimizer.get_scaled_loss(loss)
+            # for multi-gpu training, the gradient will be combined by sum, normalize the loss
+            strategy = tf.distribute.get_strategy()
+            if isinstance(strategy, tf.distribute.MirroredStrategy):
+                loss = loss / strategy.num_replicas_in_sync
+
+        elif self.framework == "torch":
+            # scale up loss for mixed precision training to avoid underflow
+            if self.model.current_optimizer.scaler is not None:
+                loss = self.model.current_optimizer.scaler.scale(loss)
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
```

Comment:
should we catch this earlier during the build method instead?

Reply:
fixed